### DISCUSSION TITLE
[@types/react] Enforcing `key` usage when needed

### DISCUSSION THREAD

**Discussion by Silic0nS0ldier** [DefinitelyTyped/DefinitelyTyped#55125] [Issues with a @types package]
Consider the following component: 

```tsx
function List<U>(props: { items: U[], renderItem: (item: U) => React.ReactNode }) {
    return <ul>{items.map(renderItem)}</ul>;
}
```

When using this unless `renderItem` is returning a fragment or element with a `key`, React will produce the warning `Each child in a list should have a unique "key" prop.`. This is a fairly common pattern, so I see this warning a lot. This isn't particularly surprising as currently (to the best of my knowledge) types don't exist to guard against this.

This warning can be "fixed" fairly easily by wrapping the result of `renderItem` in a keyed fragment, however this is a naive which can be a source of bugs (since we are giving React bad information about component identity).

A more correct solution here (which is currently possible) involves modifying the API surface, as shown below;

```tsx
function List<U>(props: { items: { key: string, value: U }[], renderItem: (item: U) => React.ReactNode }) {
    return <ul>{items.map(item => <React.Fragment key={item.key}>{renderItem(item.value}</React.Fragment>)}</ul>;
}
```

There are a few possible variations, and this works fine. However it does make for an annoying refactor, increases code complexity, and reduces performance (be it by an insignificant amount).

A cleaner solution (which is currently not possible in `@types/react`, but was validated with a POC) would express the requirement of the `key` prop in the return type. Hypothetically you might enforce the key type contract as follows;

```tsx
function List<U>(props: { items: U[], renderItem: (item: U) => React.ReactNodeWithKey }) {
    return <ul>{items.map(renderItem)}</ul>;
}
```

In theory, this "key contract" could be expanded to automatically validate arrays being passed as children including a key prop (eventually, it would make for a particularly disruptive change, best reserved for opt-in like TS strict flags).

What are people's thoughts on this?

[a comment by an automated account is not shown]

**Comment by dimitropoulos** [comment]
This would be really interesting if possible... Then, at the same time, my first reaction is what's the use-case here that isn't covered by https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-key.md already?

**Reply by Silic0nS0ldier** [reply]
The ESLint rule can only cover trivial scenarios. Taking the original example;

```tsx
function List<U>(props: { items: U[], renderItem: (item: U) => React.ReactNode }) {
    return <ul>{items.map(renderItem)}</ul>;
}
```

`renderItem` may be a prop which is supplied by a function which is defined far from its actual usage site, often in a completely different file with a non-trivial path.

---

> This would be really interesting if possible...

It is possible, I validated it prior to opening this discussion. JSX maps directly to function calls, which can have their function definitions overloaded to return a different type when `key` is defined. The hardest part would be implementing this without a breaking change (there are a lot of prop interfaces...).

An extract from my naive types experimentation follow;

```ts
// types/react/index.d.ts
interface KeyedReactElement<P = any, T extends string | JSXElementConstructor<any> = string | JSXElementConstructor<any>> extends ReactElement<P, T> {
    key: Key;
}
function createElement(
    type: "input",
    props?: InputHTMLAttributes<HTMLInputElement> & { key: Key } | null,
    ...children: ReactNode[]): KeyedReactElement;

// types/react/test/index.ts
const keyedElement: React.KeyedReactElement = React.createElement("input", { key: 'foo' });
```

**Reply by dimitropoulos** [reply]
Cool cool, didn't mean to imply it was impossible or that you hadn't done your homework in advance.

**Comment by dimitropoulos** [comment]
Do you see any downsides to doing this?  Any places it can fail or be outsmarted?

**Reply by Silic0nS0ldier** [reply]
A potential downside I can see is clarity around `key` prop contract violation. That is, what the error looks like when `key` is required but has not been supplied. It may not be clear given its the return value of `createElement` (and the new JSX transform) which we are enforcing the type check against. More investigation is needed on this front.

Should this functionality (or something like it) ship, there is potential for anti-patterns to emerge where elements with a `key` prop aren't annotated as having that prop. Especially early on, and especially with component libraries where types are authored separately. The anti-pattern in question is wrapping elements with another element in order to add a `key` prop which isn't believed to be present.

As for places where this approach can fail or be outsmarted, nothing comes to mind at present. Testing out this pattern might help, provided a large enough sample set can be built.
